File: Go/src/scraper.go
package main

import (
	"fmt"
	"log"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/PuerkitoBio/goquery"
)

type symbolsRatesType = map[string]float64
type datesSymbolsRatesType = map[time.Time]symbolsRatesType

var symbolsOrdered []string
var symbols = make(map[string]bool)
var datesSymbolsRates = make(datesSymbolsRatesType)
var goRoutineWaitGroup = sync.WaitGroup{}

func scrapeConcurrently(dates []time.Time, progressCallback func()) ([]string, datesSymbolsRatesType) {
	for _, date := range dates {
		goRoutineWaitGroup.Add(1)
		go request(date, progressCallback)
	}
	goRoutineWaitGroup.Wait()
	return symbolsOrdered, datesSymbolsRates
}

func request(date time.Time, progressCallback func()) {
	urlParams := fmt.Sprintf("?vYear=%v&vMonth=%v&vDay=%v", date.Year(), int(date.Month()), date.Day())
	url := "https://www.mongolbank.mn/dblistofficialdailyrate.aspx" + urlParams

	response, err := http.Get(url)
	if err != nil {
		log.Fatal(err)
	}
	defer response.Body.Close()
	if response.StatusCode != 200 {
		log.Fatalf("status code error: %d %s", response.StatusCode, response.Status)
	}
	doc, err := goquery.NewDocumentFromReader(response.Body)
	if err != nil {
		log.Fatal(err)
	}

	parse(date, doc)
	progressCallback()
	goRoutineWaitGroup.Done()
}

func parse(date time.Time, doc *goquery.Document) {
	spanIDPrefix := "ContentPlaceHolder1_lbl"

	elements := doc.Find(fmt.Sprintf(".uk-comment-list span[id^=%v]", spanIDPrefix))

	var symbolsRates = make(symbolsRatesType)

	elements.Each(func(_ int, element *goquery.Selection) {
		spanId, _ := element.Attr("id")
		symbol := strings.ReplaceAll(spanId, spanIDPrefix, "")
		if value := symbols[symbol]; !value {
			symbols[symbol] = true
			symbolsOrdered = append(symbolsOrdered, symbol)
		}
		if element.Text() != "" {
			rate, err := strconv.ParseFloat(strings.ReplaceAll(element.Text(), ",", ""), 64)
			if err == nil {
				symbolsRates[symbol] = rate
			}
		}
	})
	datesSymbolsRates[date] = symbolsRates
}

File: Python/src/scraper.py
import asyncio
import aiohttp
import bs4


class ScrapeConcurrently:
    symbols_ordered = []
    dates_symbols_rates = {}
    session = None
    progress_callback = None

    def __init__(self, dates, progress_callback):
        self.progress_callback = progress_callback
        asyncio.run(self.scrape_concurrently(dates))

    async def scrape_concurrently(self, dates):
        async with aiohttp.ClientSession() as session:
            self.session = session
            await asyncio.gather(*[self.request(date) for date in dates])

    async def request(self, date):
        url_params = f"?vYear={date.year}&vMonth={date.month}&vDay={date.day}"
        url = "https://www.mongolbank.mn/dblistofficialdailyrate.aspx" + url_params

        async with self.session.get(url) as response:
            doc = await response.text()
            self.parse(date, doc)
            self.progress_callback()

    def parse(self, date, doc):
        span_id_prefix = "ContentPlaceHolder1_lbl"
        elements = bs4.BeautifulSoup(doc, "html.parser").select(
            f".uk-comment-list span[id^={span_id_prefix}]"
        )
        symbols_rates = {}
        for element in elements:
            symbol = element.get("id")[len(span_id_prefix) :]
            if symbol not in self.symbols_ordered:
                self.symbols_ordered.append(symbol)
            rate_text = element.get_text()
            if rate_text != "":
                rate = float(rate_text.replace(",", ""))
                symbols_rates.update({symbol: rate})

        self.dates_symbols_rates.update({date: symbols_rates})
